src/services/WebSocketPolyfill.js

```diff
@@ -124,18 +123,21 @@ export default function initWebSocketPolyfill(syncService, fileId, initialSessio
 		}
 
 		#sendRemainingSteps() {
-			if (this.#queue.length) {
+			if (queue.length) {
+				let outbox = []
 				return syncService.sendStepsNow(() => {
+					outbox = [...queue]
 					const data = {
 						steps: this.#steps,
 						awareness: this.#awareness,
 						version: this.#version,
 					}
-					this.#queue = []
+					queue = []
 					logger.debug('sending final steps ', data)
 					return data
 				})?.catch(err => {
 					logger.error(err)
+					queue = [...outbox, ...queue]
```

Comment:
Interesting, this is an additional fix, right? It ensures that steps get reapplied in case of caught errors. Let's hope that this and the persistent queue solve synchronization issues :pray:

Reply:
Yes... that's the idea. Found some additional issues with this approach:
* As julius pointed out the queue variable get's overwritten instead of changing the underlying array.
* The sending is async so between the clearing of the queue and refilling it it's empty for some time. Fixed that by using the opposite approach - everything stays in the queue until it was successfully send.